pyamg/amg_core/ruge_stuben.h

```diff
@@ -956,345 +943,1339 @@ void cr_helper(const I A_rowptr[],  const int A_rowptr_size,
+            T denominator = 0.0;
+
+            // Start by summing entire row of A
+            for (I mm = Ap[i]; mm < Ap[i+1]; mm++) {
+                denominator += Ax[mm];
+            }
+
+            // Then subtract off the strong connections so that you are left with 
+            // denominator = a_ii + sum_{m in weak connections} a_im
+            for (I mm = Sp[i]; mm < Sp[i+1]; mm++) {
+                I this_point = Sj[mm];
+                if ( this_point != i ) {
+                    denominator -= Sx[mm]; // making sure to leave the diagonal entry in there
+                }
+                // Subtract off distance-two strong C connections that are also distance-one weak
+                // connections (i.e. not yet removed from the weak connections in denominator)
+                if (splitting[this_point] == F_NODE && (this_point != i)) {
+
+                    for (I ff = Sp[this_point]; ff < Sp[this_point+1]; ff++) {
+                        I d2_point = Sj[ff];
+
+                        // Strong C-connections to strong F-connections (distance two C connections)
+                        if (splitting[d2_point] == C_NODE) {
+
+                            // Add connection a_kl if present in matrix (search over kth row in A
+                            // for connection). Only add if sign of a_kl does not equal sign of a_kk
+                            for (I search_ind = Ap[i]; search_ind < Ap[i+1]; search_ind++) {
+                                if (Aj[search_ind] == d2_point) {
+                                    denominator -= Ax[search_ind];
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Sum over strongly connected fine points for outer denominator
+            for (I kk = Sp[i]; kk < Sp[i+1]; kk++) {
+                if ( (splitting[Sj[kk]] == F_NODE) && (Sj[kk] != i) ) {
+
+                    // Get column k and value a_ik
+                    I k = Sj[kk];
+                    T a_ik = Sx[kk];
+
+                    // Get a_ki (have to search over k'th row in A for connection a_ki)
+                    T a_ki = 0;
+                    T a_kk = 0;
+                    for (I search_ind = Ap[k]; search_ind < Ap[k+1]; search_ind++) {
+                        if (Aj[search_ind] == i) {
+                            a_ki = Ax[search_ind];
+                        }
+                        else if (Aj[search_ind] == k) {
+                            a_kk = Ax[search_ind];
+                        }
+                    }
+
+                    // If sign of a_ki matches sign of a_kk, ignore a_ki in sum
+                    // (i.e. leave as a_ki = 0)
+                    if (signof(a_ki) == signof(a_kk)) {
+                        a_ki = 0;
+                    }
+
+                    // If a_ki == 0, then we don't need to do any more work, otherwise
```

Comment:
I also wonder if we can remove the duplicate code between ext and ext+i.  We could add a boolean flag "plusi", and then change the below if statement to `if( plusi && (std:abs(a_ki) > 1e-16) )`
[ Note, I'm not adding any comments to the ext interp routine, in hopes they can be combined ]